File: UserServer/src/main/java/cc/nobrain/dev/userserver/domain/notice/entity/NoticeFile.java
package cc.nobrain.dev.userserver.domain.notice.entity;

import cc.nobrain.dev.userserver.domain.base.entity.File;
import jakarta.persistence.DiscriminatorValue;
import jakarta.persistence.Entity;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import lombok.Getter;
import org.hibernate.annotations.DynamicUpdate;

import java.util.Objects;

@Entity
@DynamicUpdate
@DiscriminatorValue("NOTICE_FILE")
@Getter
public class NoticeFile extends File {

    @ManyToOne
    @JoinColumn(name = "owner_index")
    private Notice ownerIndex;

    @Override
    public <T> void setRelation(T ownerEntity) {
        if (!(ownerEntity instanceof Notice owner) || Objects.isNull(ownerEntity)) {
            throw new IllegalArgumentException("Invalid owner entity");
        }
        this.ownerIndex = owner;
        owner.getAttachedFiles().add(this);
    }
}

File: UserServer/src/main/kotlin/cc/nobrain/dev/userserver/domain/notice/entity/NoticeFile.kt
package cc.nobrain.dev.userserver.domain.notice.entity;

import cc.nobrain.dev.userserver.domain.base.entity.File;
import jakarta.persistence.*
import org.hibernate.annotations.DynamicUpdate;

import java.util.Objects;

@Entity
@DynamicUpdate
@DiscriminatorValue("NOTICE_FILE")
class NoticeFile(
    @ManyToOne
    @JoinColumn(name = "ownerIndex", foreignKey = ForeignKey(name = "fk_ownerIndex", value = ConstraintMode.NO_CONSTRAINT))
    var ownerIndex: Notice? = null
) : File() {

    override fun <T> setRelation(ownerEntity: T) {
        if (ownerEntity !is Notice || ownerEntity == null) {
            throw IllegalArgumentException("Invalid owner entity")
        }
        this.ownerIndex = ownerEntity
        ownerEntity.attachments.add(this)
    }
}
